packages/grpc-js/src/server-call.ts

```diff
@@ -146,3 +252,213 @@ export type Handler<RequestType, ResponseType> = {
+      const metadata = this.status.metadata;
+
+      if (metadata) {
+        trailersToSend =
+            Object.assign(trailersToSend, metadata.toHttp2Headers());
+      }
+
+      this.stream.sendTrailers(trailersToSend);
+    });
+  }
+
+  receiveMetadata(headers: http2.IncomingHttpHeaders) {
+    const metadata = Metadata.fromHttp2Headers(headers);
+
+    // TODO(cjihrig): Receive compression metadata.
+
+    const timeoutHeader = metadata.get(GRPC_TIMEOUT_HEADER);
+
+    if (timeoutHeader.length > 0) {
+      const match = timeoutHeader[0].toString().match(DEADLINE_REGEX);
+
+      if (match === null) {
+        this.sendError(
+            new Error('Invalid deadline') as ServiceError, Status.OUT_OF_RANGE);
+        return;
+      }
+
+      const timeout = (+match[1] * deadlineUnitsToMs[match[2]]) | 0;
+
+      this.deadline = setTimeout(handleExpiredDeadline, timeout, this);
+      metadata.remove(GRPC_TIMEOUT_HEADER);
+    }
+
+    return metadata;
+  }
+
+  receiveUnaryMessage(): Promise<RequestType> {
+    return new Promise((resolve, reject) => {
+      const stream = this.stream;
+      const chunks: Buffer[] = [];
+      let totalLength = 0;
+
+      stream.on('data', (data: Buffer) => {
+        chunks.push(data);
+        totalLength += data.byteLength;
+      });
+
+      stream.once('end', async () => {
+        try {
+          const requestBytes = Buffer.concat(chunks, totalLength);
+
+          resolve(await this.deserializeMessage(requestBytes));
+        } catch (err) {
+          this.sendError(err, Status.INTERNAL);
+          resolve();
+        }
+      });
+    });
+  }
+
+  serializeMessage(value: ResponseType) {
+    const handler = this.handler as Handler<RequestType, ResponseType>;
+    const messageBuffer = handler.serialize(value);
+
+    // TODO(cjihrig): Call compression aware serializeMessage().
+    const byteLength = messageBuffer.byteLength;
+    const output = Buffer.allocUnsafe(byteLength + 5);
+    output.writeUInt8(0, 0);
+    output.writeUInt32BE(byteLength, 1);
+    messageBuffer.copy(output, 5);
+    return output;
+  }
+
+  async deserializeMessage(bytes: Buffer) {
+    const handler = this.handler as Handler<RequestType, ResponseType>;
+    // TODO(cjihrig): Call compression aware deserializeMessage().
+    const receivedMessage = bytes.slice(5);
+
+    return handler.deserialize(receivedMessage);
+  }
+
+  async sendUnaryMessage(
+      err: ServiceError|null, value: ResponseType|null, metadata?: Metadata,
+      flags?: number) {
+    if (err) {
+      if (metadata) {
+        err.metadata = metadata;
+      }
+
+      this.sendError(err);
+      return;
+    }
+
+    try {
+      const response = await this.serializeMessage(value as ResponseType);
+
+      if (metadata) {
+        this.status.metadata = metadata;
+      }
+
+      this.end(response);
+    } catch (err) {
+      this.sendError(err, Status.INTERNAL);
+    }
+  }
+
+  sendError(error: ServiceError, code = Status.UNKNOWN) {
```

Comment:
Instead of this I suggest the following:

```ts
sendStatus(statusObj: StatusObject) {
    if (this.cancelled === true) {
      return;
    }

     if (this.deadline !== null) {
      clearTimeout(this.deadline);
      this.deadline = null;
    }
  this.sendMetadata();
  this.stream.once('wantTrailers', () => {
      let trailersToSend = {
        [GRPC_STATUS_HEADER]: status.code,
        [GRPC_MESSAGE_HEADER]: encodeURI(status.details as string)
      };
      const metadata = status.metadata;

       if (metadata) {
        trailersToSend =
            Object.assign(trailersToSend, metadata.toHttp2Headers());
      }

       this.stream.sendTrailers(trailersToSend);
    });
  this.stream.end();
}
```

Then `end` should be removed, and the trailer code should be removed from `sendMetadata`. The code to translate a `ServiceError` to a `StatusObject` would have to be pushed up a layer, but this would significantly simplify the logic of ending a response

Reply:
Renaming this and switching to a `StatusObject` input shouldn't be too big of a deal, but what about the following:

> Then `end` should be removed

The server and bidi streams still have to signal down to the HTTP2 layer to end the call. The deadline timer is also cleared there, along with any future cleanup code that might be added. I imagine this would result in the code being duplicated, or still being contained in a function, just with a name other than `end`.

> the trailer code should be removed from `sendMetadata`

Don't we still want this behavior in places other than `sendStatus()` (for example `sendUnaryMessage()`)?

Reply:
The idea here is that the only path to `stream.end()` would be going through `sendStatus()`. I included the existing code from `end()` in my suggested code. I think a reasonable alternative would be to keep `end` with the cleanup code, but make it private and call it only from here. And the idea is that `sendUnaryMessage()` would also call `sendStatus()` at the end; it's already calling `sendError()` conditionally a couple of times, it could simplify the code to instead call `sendStatus()` unconditionally, perhaps with a default simple status.

The point of this suggestion is to ensure that any code path for ending a call goes through setting some status, and that setting the trailers goes through the same path, because it is essentially the same action.

Reply:
Thanks for the explanation. I've updated the PR with `sendStatus()`. I kept `sendError()` as a helper for translating `ServiceError`s before calling `sendStatus()`.